File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int N, M, answer = 0;
    cin >> N >> M;
    vector<int> woods(1000001);
    for (int i = 0; i < N; i++) {
        cin >> woods[i];
    }
    sort(woods.begin(), woods.begin() + N);
    int left = 1, right = woods[N - 1], mid;
    while (left <= right) {
        long long mod = 0;
        mid = (left + right) / 2;
        for (int i = 0; i < N; i++) {
            if (woods[i] > mid) mod += woods[i] - mid;
        }
        if (mod >= M) {
            answer = max(answer, mid);
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    cout << answer;

}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static ArrayList<Integer> woods;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());
        N = Integer.parseInt(tokenizer.nextToken());
        M = Integer.parseInt(tokenizer.nextToken());
        woods = new ArrayList<>();
        tokenizer = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            woods.add(Integer.parseInt(tokenizer.nextToken()));
        }
        long result  = solution();
        bw.write(result + "");
        bw.flush();
        bw.close();
        br.close();
    }

    private static long solution() {
        long left = 0, right = Collections.max(woods);
        long mid;
        while (left <= right) {
            mid = (left + right) / 2;
            long sum = 0;
            for (long wood : woods) {
                if (wood - mid > 0) {
                    sum += wood - mid;
                }
            }
            if (sum >= M) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }
}